### PR TITLE
[core modules] Add `ToLBRExchangeRateUpdateEvent` to the `Libra` module

### PR DESCRIPTION
**Pull request by tzakian**
Adds an event for when the `to_lbr_exchange_rate` for a currency is updated.

### CODE REVIEW

File: types/src/account_config/events/exchange_rate_update.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+use crate::account_config::LIBRA_MODULE_NAME;
+use anyhow::Result;
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    move_resource::MoveResource,
+};
+use serde::{Deserialize, Serialize};
+
+/// Struct that represents a MintEvent.
```
Comment (sblackshear) [inline types/src/account_config/events/exchange_rate_update.rs:12]:
```suggestion
/// Struct that represents a ToLBRExchangeRateUpdateEvent.
```

File: types/src/account_config/events/exchange_rate_update.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    new_to_lbr_exchange_rate: u64,
+}
+
+impl ToLBRExchangeRateUpdateEvent {
+    /// Exchange rates are represented as 32|32 fixed-point numbers
+    pub fn new_to_lbr_exchange_rate(&self) -> f32 {
+        (self.new_to_lbr_exchange_rate as f32) / 2f32.powf(32f32)
+    }
+
+    /// Return the code for the currency that was minted
```
Comment (sblackshear) [inline types/src/account_config/events/exchange_rate_update.rs:25]:
```suggestion
    /// Return the code for the currency that was updated
```

File: language/stdlib/modules/Libra.move [hunk 3]
```diff
@@ -551,7 +564,8 @@ module Libra {
             mint_events: Event::new_event_handle<MintEvent>(account),
             burn_events: Event::new_event_handle<BurnEvent>(account),
             preburn_events: Event::new_event_handle<PreburnEvent>(account),
-            cancel_burn_events: Event::new_event_handle<CancelBurnEvent>(account)
+            cancel_burn_events: Event::new_event_handle<CancelBurnEvent>(account),
+            exchange_rate_update_events: Event::new_event_handle<ToLBRExchangeRateUpdateEvent>(account)
```
Comment (sblackshear) [inline language/stdlib/modules/Libra.move:568]: Hopefully our tests will catch this, but just wanted to double-check that this isn't changing any important event keys for handles created later in genesis...

File: language/stdlib/modules/Libra.move [hunk 4]
```diff
@@ -630,6 +644,14 @@ module Libra {
         assert_assoc_and_currency<FromCoinType>(account);
         let currency_info = borrow_global_mut<CurrencyInfo<FromCoinType>>(CoreAddresses::CURRENCY_INFO_ADDRESS());
         currency_info.to_lbr_exchange_rate = lbr_exchange_rate;
+        Event::emit_event(
+            &mut currency_info.exchange_rate_update_events,
+            ToLBRExchangeRateUpdateEvent {
+                currency_code: *&currency_info.currency_code,
```
Comment (sblackshear) [inline language/stdlib/modules/Libra.move:650]: @tnowacki: does it work to write `copy currency_info.currency_code` here (and if it does, is there a reason to prefer one of `copy` and `*&`?

File: types/src/account_config/events/exchange_rate_update.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ToLBRExchangeRateUpdateEvent {
+    currency_code: Identifier,
+    new_to_lbr_exchange_rate: u64,
+}
+
+impl ToLBRExchangeRateUpdateEvent {
+    /// Exchange rates are represented as 32|32 fixed-point numbers
+    pub fn new_to_lbr_exchange_rate(&self) -> f32 {
+        (self.new_to_lbr_exchange_rate as f32) / 2f32.powf(32f32)
```
Comment (sblackshear) [inline types/src/account_config/events/exchange_rate_update.rs:22]: This requires more explanation (at least for me...)

File: language/stdlib/modules/Libra.move (hunk 4 above)
Comment (tzakian) [reply language/stdlib/modules/Libra.move:650]: I actually wound up writing `copy currency_info.currency_code` here at first, but the parser was unhappy with that. I didn't try `copy &currency_info.currency_code`, but I'm guessing the compiler wants to see ` &` in front of a field access to know that it's grabbing a reference to that field.

File: language/stdlib/modules/Libra.move (hunk 3 above)
Comment (tzakian) [reply language/stdlib/modules/Libra.move:568]: Our tests do indeed catch this :) (see the change to `block_metadata.rs`).

File: types/src/account_config/events/exchange_rate_update.rs (hunk 5 above)
Comment (tzakian) [reply types/src/account_config/events/exchange_rate_update.rs:22]: ahhh yes.... I got this from here:
https://medium.com/incredible-coder/converting-fixed-point-to-floating-point-format-and-vice-versa-6cbc0e32544e#:~:text=Converting%20from%20fixed%2Dpoint%20to%20floating%2Dpoint&text=Divide%20the%20number%20by%202,to%20the%20power%20of%20n).

and also tested it (valid up-to rounding errors): 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2456fe91c9c3034233555f4aa4fe7e08

**Comment by n4ss** [comment]
That looks great!

**Pull request closed without merging**